File: common/hogvm/python/stl/date.py
import re
import datetime
from typing import Optional

import pytz

from common.hogvm.python.objects import is_hog_date, is_hog_datetime


def to_hog_date(year: int, month: int, day: int):
    return {
        "__hogDate__": True,
        "year": year,
        "month": month,
        "day": day,
    }


def to_hog_datetime(timestamp: int | float | dict, zone: Optional[str] = None):
    if isinstance(timestamp, dict) and is_hog_date(timestamp):
        dt = datetime.datetime(
            year=timestamp["year"], month=timestamp["month"], day=timestamp["day"], tzinfo=pytz.timezone(zone or "UTC")
        )
        return {
            "__hogDateTime__": True,
            "dt": dt.timestamp(),
            "zone": (dt.tzinfo.tzname(None) if dt.tzinfo else None) or "UTC",
        }
    return {
        "__hogDateTime__": True,
        "dt": timestamp,
        "zone": zone or "UTC",
    }


# Exported functions


def now(zone: Optional[str] = None):
    return to_hog_datetime(datetime.datetime.now().timestamp(), zone)


def toUnixTimestamp(date, timezone: Optional[str] = None):
    if isinstance(date, dict) and is_hog_datetime(date):
        return date["dt"]
    if isinstance(date, dict) and is_hog_date(date):
        return datetime.datetime(
            year=date["year"], month=date["month"], day=date["day"], tzinfo=pytz.timezone(timezone or "UTC")
        ).timestamp()

    # A naive string is anchored to `timezone` (UTC when unset), matching Rust's
    # `naive_to_seconds(s, zone)` and TypeScript's `fromISO(input, {zone})`. It used to be parsed
    # naive and then `.astimezone`d, which is a no-op for aware input and resolves naive input in the
    # *host's* timezone — so the answer depended on where the process happened to be running.
    parsed = _parse_date_like(date, timezone)
    if parsed is None:
        raise ValueError(f"Could not parse date: {date}")
    return parsed.timestamp()


def fromUnixTimestamp(timestamp: int | float):
    return to_hog_datetime(timestamp)


def toUnixTimestampMilli(date, timezone: Optional[str] = None):
    return int(toUnixTimestamp(date, timezone) * 1000)


def fromUnixTimestampMilli(timestamp: int):
    return fromUnixTimestamp(float(timestamp) / 1000.0)


def toTimeZone(date: dict, timezone: str):
    if not is_hog_datetime(date):
        raise ValueError("Expected a DateTime")
    return {
        **date,
        "zone": timezone,
    }


# [0-9] and never \d: Python's re treats \d as any Unicode decimal digit, which int() then happily
# parses — so e.g. an all-Arabic-Indic-digit date was a valid instant here and rejected by the
# other two VMs. See the canonical grammar in the Rust STL.
DATE_LIKE = re.compile(
    r"""^([0-9]{4})-([0-9]{2})-([0-9]{2})
        (?:
            [Tt\ ]([01][0-9]|2[0-3]):([0-5][0-9])
            (?::([0-5][0-9])(?:[.,]([0-9]{1,9}))?)?
            (Z|z|[+-](?:[01][0-9]|2[0-3])(?::?[0-5][0-9])?)?
        )?\Z""",
    re.VERBOSE,
)


def _parse_date_like(input: str, timezone: Optional[str] = None) -> Optional[datetime.datetime]:
    """An aware ``datetime`` for a string matching the shared date-like grammar, else None.

    The canonical copy of the grammar lives above ``parse_datetime_to_seconds`` in
    ``rust/common/hogvm/src/stl.rs``; ``common/hogvm/typescript/src/stl/date.ts`` (``parseDateLike``)
    is the third implementation. Change all three together.

    Deliberately not ``datetime.fromisoformat``, which accepts the compact ``20240101`` and ISO-week
    ``2024-W05`` forms that neither of the other two VMs do. More importantly, ``fromisoformat``
    returns a *naive* datetime for input carrying no zone, and ``naive.timestamp()`` resolves it in
    the **host's** timezone — so the Python VM's answer for ``toDateTime('2026-07-01')`` depended on
    where it happened to be running, and disagreed with TypeScript and Rust (both UTC-anchored) by
    the local UTC offset. ``tzinfo`` is always set here, so the result is host-independent.

    ``timezone`` applies only to input carrying no zone of its own; an explicit offset or ``Z`` pins
    the absolute instant regardless.
    """
    match = DATE_LIKE.match(input.strip())
    if not match:
        return None
    year, month, day, hour, minute, second, fraction, offset = match.groups()
    # Sub-millisecond digits are truncated, not rounded, to match luxon and Rust's `timestamp_millis`.
    # Python's `datetime` keeps microseconds, which surfaced as a `result_mismatch` against the Node
    # baseline (see `datetime_to_seconds` in the Rust STL).
    microsecond = int(fraction[:3].ljust(3, "0")) * 1000 if fraction else 0

    try:
        if offset in ("Z", "z"):
            tzinfo: datetime.tzinfo = datetime.UTC
        elif offset:
            digits = offset[1:].replace(":", "")
            delta = datetime.timedelta(hours=int(digits[:2]), minutes=int(digits[2:4] or 0))
            tzinfo = datetime.timezone(-delta if offset[0] == "-" else delta)
        else:
            tzinfo = pytz.timezone(timezone or "UTC")

        naive = datetime.datetime(
            int(year), int(month), int(day), int(hour or 0), int(minute or 0), int(second or 0), microsecond
        )
    except ValueError:
        # Out-of-range calendar date (2024-02-30) or year 0. The tz construction is inside the `try`
        # deliberately: an offset the regex admits but `datetime.timezone` rejects must return None
        # like any other parse failure — this function is called from `unify_comparison_types` on
        # every comparison opcode, where a bare ValueError would escape the VM entirely rather than
        # leaving the operands uncoerced. (An unknown zone *name* still raises — pytz's
        # UnknownTimeZoneError is a KeyError — matching the natives' pre-existing behavior; only
        # they pass a zone, never the comparison path.)
        return None
    # `localize` rather than `tzinfo=` — pytz zones carry a historical LMT offset that only
    # `localize` resolves to the right one for the date in question. `is_dst=True` picks the first
    # of an ambiguous pair during a DST fold, matching Rust's `LocalResult::Ambiguous(dt, _)` and
    # luxon; pytz's default (`is_dst=False`) would pick the second and land an hour off.
    localize = getattr(tzinfo, "localize", None)
    return localize(naive, is_dst=True) if localize else naive.replace(tzinfo=tzinfo)


def toDate(input):
    if isinstance(input, int) or isinstance(input, float):
        dt = datetime.datetime.fromtimestamp(input)
    else:
        parsed = _parse_date_like(input)
        if parsed is None:
            raise ValueError(f"Could not parse date: {input}")
        dt = parsed
    return {
        "__hogDate__": True,
        "year": dt.year,
        "month": dt.month,
        "day": dt.day,
    }


def toDateTime(input):
    if isinstance(input, int) or isinstance(input, float):
        dt = float(input)
    else:
        parsed = _parse_date_like(input)
        if parsed is None:
            raise ValueError(f"Could not parse date: {input}")
        dt = parsed.timestamp()
    return {
        "__hogDateTime__": True,
        "dt": dt,
        "zone": "UTC",
    }


def date_string_to_seconds(input: str) -> Optional[float]:
    """Epoch seconds for a date-like string, parsed the same way `toDateTime` would, else None."""
    parsed = _parse_date_like(input)
    return parsed.timestamp() if parsed else None


# From ClickHouse to Python
token_translations = {
    "a": "%a",
    "b": "%b",
    "c": "%m",
    "C": "%y",
    "d": "%d",
    "D": "%m/%d/%y",
    "e": "%d",
    "f": "%f",
    "F": "%Y-%m-%d",
    "g": "%y",
    "G": "%Y",
    "h": "%I",
    "H": "%H",
    "i": "%M",
    "I": "%I",
    "j": "%j",
    "k": "%H",
    "l": "%I",
    "m": "%m",
    "M": "%B",
    "n": "\n",
    "p": "%p",
    # 'Q': '%Q',
    "r": "%I:%M %p",
    "R": "%H:%M",
    "s": "%S",
    "S": "%S",
    "t": "\t",
    "T": "%H:%M:%S",
    "u": "%u",
    "V": "%V",
    "w": "%w",
    "W": "%A",
    "y": "%y",
    "Y": "%Y",
    "z": "%z",
    "%": "%%",
}


def formatDateTime(input: dict, format: str, zone: Optional[str] = None) -> str:
    if not is_hog_datetime(input):
        raise ValueError("Expected a DateTime")
    format_string = ""
    acc = ""
    i = 0
    while i < len(format):
        if format[i] == "%":
            if acc:
                format_string += acc
                acc = ""
            i += 1
            if i < len(format) and format[i] in token_translations:
                format_string += token_translations[format[i]]
        else:
            acc += format[i]
        i += 1
    if acc:
        format_string += acc
    return datetime.datetime.fromtimestamp(input["dt"], pytz.timezone(zone or input["zone"])).strftime(format_string)

File: common/hogvm/typescript/src/stl/date.ts
import { DateTime } from 'luxon'

import { isHogDate, isHogDateTime } from '../objects'
import { HogDate, HogDateTime } from '../types'

export function toHogDate(year: number, month: number, day: number): HogDate {
    return {
        __hogDate__: true,
        year: year,
        month: month,
        day: day,
    }
}

export function toHogDateTime(timestamp: number | HogDate, zone?: string): HogDateTime {
    if (isHogDate(timestamp)) {
        const dateTime = DateTime.fromObject(
            {
                year: timestamp.year,
                month: timestamp.month,
                day: timestamp.day,
            },
            { zone: zone || 'UTC' }
        )
        return {
            __hogDateTime__: true,
            dt: dateTime.toSeconds(),
            zone: dateTime.zoneName || 'UTC',
        }
    }
    return {
        __hogDateTime__: true,
        dt: timestamp,
        zone: zone || 'UTC',
    }
}

// EXPORTED STL functions

export function now(zone?: string): HogDateTime {
    return toHogDateTime(Date.now() / 1000, zone)
}

export function toUnixTimestamp(input: HogDateTime | HogDate | string, zone?: string): number {
    if (isHogDateTime(input)) {
        return input.dt
    }
    if (isHogDate(input)) {
        return toHogDateTime(input).dt
    }
    return (parseDateLike(input, zone) ?? INVALID).toSeconds()
}

export function fromUnixTimestamp(input: number): HogDateTime {
    return toHogDateTime(input)
}

export function toUnixTimestampMilli(input: HogDateTime | HogDate | string, zone?: string): number {
    return toUnixTimestamp(input, zone) * 1000
}

export function fromUnixTimestampMilli(input: number): HogDateTime {
    return toHogDateTime(input / 1000)
}

export function toTimeZone(input: HogDateTime, zone: string): HogDateTime | HogDate {
    if (!isHogDateTime(input)) {
        throw new Error('Expected a DateTime')
    }
    return { ...input, zone }
}

/**
 * The shared "date-like string" grammar, implemented identically by all three HogVMs. The canonical
 * copy of the spec lives above `parse_datetime_to_seconds` in `rust/common/hogvm/src/stl.rs`;
 * `common/hogvm/python/stl/date.py` (`_parse_date_like`) is the third. Change all three together.
 *
 *     input := WS* date ( SEP time zone? )? WS*
 *     date  := YYYY "-" MM "-" DD              # extended format only, YYYY >= 0001
 *     SEP   := "T" | "t" | " "
 *     time  := HH ":" MM ( ":" SS frac? )?     # HH <= 23; a fraction requires seconds
 *     frac  := ("." | ",") DIGIT{1,9}          # truncated to milliseconds
 *     zone  := "Z" | "z" | ("+"|"-") HH ( ":"? MM )?   # offset HH <= 23, MM <= 59
 *
 * Luxon's `fromISO` is both too permissive and too strict for this. Too permissive: it accepts
 * `2024`, `2024-01`, `20240101`, `2024-W05`, `2024-001`, and a time-only `12:30` that silently
 * resolves to *today's* date — all plausible values of an ordinary string property, and this
 * coercion is implicit, so a false positive changes the meaning of a comparison the filter author
 * never marked as a date comparison. Too strict: it rejects `2024-01-01 00:00:00`, the canonical
 * ClickHouse form that HogQL emits and that Python and Rust both accept — so `toDateTime` on it
 * returned `{dt: NaN}`. Validating against the grammar first fixes both directions.
 */
// [0-9] to stay byte-for-byte parallel with the Rust and Python patterns, where \d is
// Unicode-aware and had to be banned (JS's \d is already ASCII-only).
const DATE_LIKE =
    /^([0-9]{4})(-[0-9]{2}-[0-9]{2})(?:[Tt ]((?:[01][0-9]|2[0-3]):[0-5][0-9])(?::([0-5][0-9])(?:[.,]([0-9]{1,9}))?)?(Z|z|[+-](?:[01][0-9]|2[0-3])(?::?[0-5][0-9])?)?)?$/

/** Luxon `DateTime` for a string matching the shared grammar, else null. `zone` applies only to input carrying no zone of its own. */
export function parseDateLike(input: string, zone?: string): DateTime | null {
    // Callers reach here with whatever an event property held — `toUnixTimestamp(event.properties.x)`
    // passes its argument through unchecked — so a non-string must return null, not throw on `.trim`.
    if (typeof input !== 'string') {
        return null
    }
    const match = DATE_LIKE.exec(input.trim())
    if (!match) {
        return null
    }
    const [, year, monthDay, hourMinute, second, fraction, offset] = match
    // Year 0 is valid to luxon and chrono but not to Python's `datetime`; excluded so the three
    // accept-sets stay identical.
    if (year === '0000') {
        return null
    }
    // Sub-millisecond digits are truncated, not rounded, to match luxon's own precision and Rust's
    // `timestamp_millis`. Keeping microseconds (as Python's `datetime` does) surfaced as a
    // `result_mismatch` against the Node baseline.
    const millis = fraction ? `.${fraction.slice(0, 3).padEnd(3, '0')}` : ''
    // Normalize to strict ISO so luxon's parser accepts it: `T` separator, uppercase `Z`.
    const time = hourMinute ? `T${hourMinute}${second ? `:${second}${millis}` : ''}${offset?.toUpperCase() ?? ''}` : ''
    const dt = DateTime.fromISO(`${year}${monthDay}${time}`, { zone: zone || 'UTC' })
    return dt.isValid ? dt : null
}

// NOTE: unparseable input keeps each VM's *existing* failure mode — here an invalid luxon DateTime
// (so `.year`/`.toSeconds()` are NaN, as before), while Python raises and Rust errors into a null.
// Converging those three is a separate change; this only converges *what parses*.
const INVALID = DateTime.invalid('not a date-like string')

export function toDate(input: string | number): HogDate {
    // Previously `DateTime.fromISO(input)` with no zone, i.e. the *system* zone — same class of
    // host-dependent bug as Python's, and a day off from `toDateTime` for offsets west of UTC.
    const dt = typeof input === 'number' ? DateTime.fromSeconds(input) : (parseDateLike(input) ?? INVALID)
    return {
        __hogDate__: true,
        year: dt.year,
        month: dt.month,
        day: dt.day,
    }
}

export function toDateTime(input: string | number, zone?: string): HogDateTime {
    const dt = typeof input === 'number' ? input : (parseDateLike(input, zone) ?? INVALID).toSeconds()
    return {
        __hogDateTime__: true,
        dt: dt,
        zone: zone || 'UTC',
    }
}

/** Epoch seconds for a date-like string, parsed the same way `toDateTime` would, else null. */
export function dateStringToSeconds(input: string): number | null {
    return parseDateLike(input)?.toSeconds() ?? null
}

/** Convert from ClickHouse format string to Luxon format string */
const tokenTranslations: Record<string, string> = {
    a: 'EEE',
    b: 'MMM',
    c: 'MM',
    C: 'yy',
    d: 'dd',
    D: 'MM/dd/yy',
    e: 'd',
    f: 'SSS',
    F: 'yyyy-MM-dd',
    g: 'yy',
    G: 'yyyy',
    h: 'hh',
    H: 'HH',
    i: 'mm',
    I: 'hh',
    j: 'ooo',
    k: 'HH',
    l: 'hh',
    m: 'MM',
    M: 'MMMM',
    n: '\n',
    p: 'a',
    Q: 'q',
    r: 'hh:mm a',
    R: 'HH:mm',
    s: 'ss',
    S: 'ss',
    t: '\t',
    T: 'HH:mm:ss',
    u: 'E',
    V: 'WW',
    w: 'E',
    W: 'EEEE',
    y: 'yy',
    Y: 'yyyy',
    z: 'ZZZ',
    '%': '%',
}
export function formatDateTime(input: any, format: string, zone?: string): string {
    if (!isHogDateTime(input)) {
        throw new Error('Expected a DateTime')
    }
    if (!format) {
        throw new Error('formatDateTime requires at least 2 arguments')
    }
    let formatString = ''
    let acc = ''
    for (let i = 0; i < format.length; i++) {
        if (format[i] === '%') {
            if (acc.length > 0) {
                formatString += `'${acc}'`
                acc = ''
            }
            i += 1
            if (i < format.length && tokenTranslations[format[i]]) {
                formatString += tokenTranslations[format[i]]
            }
        } else {
            acc += format[i]
        }
    }
    if (acc.length > 0) {
        formatString += `'${acc}'`
    }
    return DateTime.fromSeconds(input.dt, { zone: zone || input.zone }).toFormat(formatString)
}
